hail/src/main/scala/is/hail/expr/ir/Requiredness.scala

```diff
@@ -96,6 +96,14 @@ class Requiredness(val usesAndDefs: UsesAndDefs, ctx: ExecuteContext) {
 
   def addBindingRelations(node: BaseIR): Unit = {
     val refMap = usesAndDefs.uses(node).toArray.groupBy(_.t.name).mapValues(_.asInstanceOf[Array[RefEquality[BaseIR]]])
+    def addDeflessBinding(name: String, d: IR, req: BaseTypeWithRequiredness): Unit = {
+      if (refMap.contains(name)) {
+        val uses = refMap(name)
+        uses.foreach { u => defs.bind(u, Array(req)) }
+        dependents.getOrElseUpdate(d, mutable.Set[RefEquality[BaseIR]]()) ++= uses
+      }
+    }
```

Comment:
I can't figure out how this works.  I know what needs to happen: the ShuffeWrite needs need to determine the requiredness of the TShuffle, which in turn determines the requiredness of the read and partition bounds nodes.  I'm still a little confused by the implementation of Requiredness but I'll keep reading it over.

Reply:
@catoverdrive you should probably review the changes to this file.

Reply:
I'll keep trying to understand this, but I would also appreciate @catoverdrive's review.